python/pyspark/ml/tuning.py

```diff
@@ -536,7 +536,7 @@ def copy(self, extra=None):
         bestModel = self.bestModel.copy(extra)
         avgMetrics = self.avgMetrics
         subModels = self.subModels
-        return CrossValidatorModel(bestModel, avgMetrics, subModels)
+        return self._copyValues(CrossValidatorModel(bestModel, avgMetrics, subModels), extra=extra)
```

Comment:
Should we use `Params.copy` like `CrossValidator`?

Reply:
We can. However I just found one potential issue with using `Params.copy` (not specific to `CrossValidator`). It creates a shallow copy of `self` (i.e. the models). Hence if we run the below snippet
```python
cvModelCopied = cvModel.copy()
cvModel.avgMetrics[0] = 'foo'
assert cvModelCopied.avgMetrics[0] != 'foo'  # This will fail
```
Based on the Scala equivalent I think `avgMetrics` should be shallow copied and `subModels` should be copied with the copying actions delegated to `copy()` of each model. I will push a change to this function.

Reply:
You meant avgMetrics should be or should not be shallow copied?

Reply:
`avgMetrics` should be shallow copied, as tested in https://github.com/Louiszr/spark/blob/8ae74d000ac48d6e293feb4bc3dac31088bf6c6c/python/pyspark/ml/tests/test_tuning.py#L124-L129
I think `Params.copy` will shallow copy the `CrossValidatorModel` object and thus only copy the reference to `avgMetrics`

Reply:
Hm? I think the test makes sure it isn't shallow copy but deep copy, isn't?

Reply:
By shallow copy I mean `copy.copy()` in python, which makes re-assigning `cvModel.avgMetrics[0]` not being propagated to `cvModelCopied.avgMetrics[0]`.
I am also using `copy.deepcopy()` as the reference for deep copy. If `cvModel.avgMetrics[0]` is an class instance, then shallow copy will point to the same instance, while deep copy will create a copy of the instance.
I think here it doesn't make a difference because `avgMetrics` is a list of `float`, but in the future if it does become a list of objects then a shallow copy implementation will be sufficient to pass the test.